### ISSUE TITLE
World Corrupted; NullReference errors when trying to enter buildings

### ISSUE THREAD

**Issue by Ipiano** [LunaMultiplayer/LunaMultiplayer#330]
------------------------------------------------------------------------------

OS: Windows 10 v1909 Build 18363.535

------------------------------------------------------------------------------

KSPVersion: 1.8.1.2694 w/ no DLC

------------------------------------------------------------------------------

LMP version: 0.24.0

------------------------------------------------------------------------------

MODS: None

------------------------------------------------------------------------------

Expected behaviour: When I attempt to enter buildings at the Space Center, the game behaves as normal and allows me to enter the buildings.

------------------------------------------------------------------------------

Current behaviour: When I click on certain buildings (VAB, Spaceplane hanger, Tracking Station for sure) nothing happens. When I click on other buildings (Research Center, Astronaut Complex for sure) I enter the building, but cannot leave it. After joining the server, I cannot exit to main menu through the regular KSP UI.

------------------------------------------------------------------------------

Steps to reproduce: Unsure exactly what caused this; but we are suspicious that it is related to a number of kerbals we've had go missing on the server as per issue #328 as well as a number of ships that have gone missing; possibly due to a number of reverts during their launch. (We've seen a some funky behaviors with reverting, including the reverted ship being still on the launch pad after launching again, and reverted ships not counting as 'newly built' for the purpose of missions). Most recently, there was a mun base that just disappeared after restarting the server, leaving the flag planted by it. But if you load up the universe attached, you should experience this behavior; it does not appear to be localized to just one client.

Further evidence that this is related to funkiness around reverting/missing kerbals: The KSP.log file has a bunch of errors that Kerbals are supposed to be assigned, but then not referenced by a ship, and so they go MIA; then later on (at the end) there's NullReference errors trying to do something related to rosters. I don't know much about the KSP internals, but this sounds like an issue related to the list of Kerbonauts being screwed up. 

If possible, we'd love to recover this universe, so I plan to mess around with some of the files and see if I can get it back to a working state.

------------------------------------------------------------------------------
Attachments
* [KSP.log](https://github.com/LunaMultiplayer/LunaMultiplayer/files/4062905/KSP.log)
* [output_log.txt](https://github.com/LunaMultiplayer/LunaMultiplayer/files/4062889/output_log.txt)
* [Universe + Config.zip](https://github.com/LunaMultiplayer/LunaMultiplayer/files/4062899/Universe.%2B.Config.zip)

**Comment by Ipiano** [comment]
Update: I was able to find this other, [similar issue](https://github.com/KSP-RO/RP-0/issues/892), which pointed to [here](https://forum.kerbalspaceprogram.com/index.php?/topic/157651-cant-exit-buildings-activate-flights/), which led to [here](https://forum.kerbalspaceprogram.com/index.php?/topic/95213-ksc-buildings-lock-up-cant-click-on-anything-firing-kerbal-crashes-game/). The suggested solution there is to modify the `persistence.sfs` file for the save. 

Without digging into the LMP code, I've made the following guesses:
* That that file is generated by the server (or on the clients, from data on the server) from the files in the server's Universe directory
* The Progress Tracking and Contract sections in the .sfs file are formed from `Scenarios/ProgressTracking.txt` and `Scenarios/ContractSystem.txt`, respectively
* The crew section of the .sfs file is formed from all of the Kerbal txt files in `Kerbals`

With those assumptions, I started trying to make the suggested edits, deleting the LunaMultiplayer save on my client each time to make sure things were synced only from the server. I was unable to make this work, but here are the things I tried
* Generating a random kerbal file with appropriate data for each missing kerbal listed
* Deleting contracts with missing kerbals (There were no missing kerbals in the progress tracker)

Aside: I noticed in the progress tracker that a lot of the crew lists were just duplicates of the same set of kerbals; and having Jebediah Kerman listed 10 times as the crew didn't seem right, so I also trimming them all down to just contain one of each listed; but that doesn't appear to have helped.

List of kerbals I found to be missing; they all appear to come from 3 of the group tourist contracts
* Dillorf Kerman
* Dofen Kerman
* Edmon Kerman
* Herfurt Kerman
* Tius Kerman
* Ferden Kerman
* Shepmon Kerman
* Gilpond Kerman
* Kirzor Kerman
* Magski Kerman
* Mauwise Kerman
* Jortop Kerman
* Kertop Kerman
* Phosel Kerman

**Comment by Ipiano** [comment]
So looking through some of the other closed issues; my guess is that this is just related to the fact that it is 'difficult to sync tourists'. I'm unsure why we were getting tourist contracts, if they are disabled; perhaps those were disabled after the most recent release?

In any case, I'm guessing this will likely get closed as 'contains tourists, they're disabled, won't fix' but it'd be nice if the server could recover from this situation regardless. I guess I don't know what all that entails, or if it would be possible to get into a state like this without tourists.

**Comment by Ipiano** [comment]
Yet another update: We started over, and this happened again. Right around at the same point in the game too - Right after we landed on Minmus for the first time. I was able to recover the game this time around by removing all contracts that were not complete or active, and then adding them back in in a binary-search fashion. Narrowed the issue down to a single contract that listed an existing kerbal as a tourist, even though he was already a scientist. Guessing something similar happened last time too, but at this point it's not worth trying to recover it. Like I said, it'd be nice if the server could check for some of these consistency things to at least report them, but that's probably out of scope for this project.
